firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,1035 @@
+
+    return numpy.unique(boundary_nodes[:idx])
+
+
+@cython.wraparound(False)
+def cell_entity_masks(mesh):
+    """Compute a masking integer for each cell in the extruded mesh.
+
+    This integer indicates for each cell, which topological entities
+    in the cell are on the boundary of the domain.  If the ith bit in
+    the integer is on, that indicates that the ith entity is on the
+    boundary, meaning that the appropriate boundary mask should be
+    used to discard element tensor contributions when assembling
+    bilinear forms.
+
+    :arg mesh: the extruded mesh.
+    :returns: a tuple of section, bottom, and top masks.  The section
+        records the number of entities in each column and the offset
+        in the masking arrays for the start of each column.
+    """
+    cdef:
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] cell_closure = mesh.cell_closure
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents = mesh.layer_extents
+        numpy.ndarray[numpy.int64_t, ndim=1, mode="c"] top_mask, bottom_mask
+        numpy.ndarray[numpy.int64_t, ndim=1, mode="c"] flips
+        numpy.ndarray[PetscInt, ndim=1, mode="c"] points
+        int cell, ncell, nclosure, idx
+        PetscInt cStart, cEnd
+        PETSc.Section section
+        PetscInt c, cell_bottom, cell_top, layer, p, e, ent, point
+        PetscInt ent_top, ent_bottom
+
+    ncell = cell_closure.shape[0]
+    nclosure = cell_closure.shape[1]
+    cStart, cEnd = mesh._plex.getHeightStratum(0)
+    top_mask = numpy.zeros(numpy.sum(layer_extents[cStart:cEnd, 1] - layer_extents[cStart:cEnd, 0] - 1),
+                           dtype=numpy.int64)
+    bottom_mask = numpy.zeros_like(top_mask)
+    # We iterate over the base cell and do all the entities in the extruded cell above it,
+    # therefore we need a mapping from the standard firedrake entity ordering to this one.
+    fiat_cell = as_fiat_cell(mesh.ufl_cell())
+    points = eutils.entity_reordering(fiat_cell)
+    # Some masks incorporate others (since the closure of the entity
+    # is not just the entity itself), we'll use this to minimise the
+    # number of mask bits that are on.
+    flips = numpy.zeros(points.shape[0], dtype=numpy.int64)
+    for k, v in eutils.entity_closures(fiat_cell).items():
+        flips[k] = ~sum(2**x for x in v)
+    idx = 0
+    section = PETSc.Section().create(comm=PETSc.COMM_SELF)
+    section.setChart(0, ncell)
+    for cell in range(ncell):
+        c = cell_closure[cell, nclosure-1]
+        cell_bottom = layer_extents[c, 0]
+        cell_top = layer_extents[c, 1]
+        CHKERR(PetscSectionSetDof(section.sec, cell, cell_top - cell_bottom - 1))
+        for layer in range(cell_bottom, cell_top - 1):
+            p = 0
+            # Iteration order.  Normal FIAT ordering for the entities
+            # in the base cell.  Then, on each entity, we look at the
+            # bottom vertex of the interval, then the top vertex, then
+            # the cell.
```

Comment:
IOW, `point[p]` is the entity at the bottom of the extruded interval, `point[p+1]` is the entity at the top, and `point[p+2]` is the full extruded interval.